web/modules/custom/nidirect_prisons/src/Controller/WorldpayNotificationController.php

```diff
@@ -0,0 +1,268 @@
+<?php
+
+namespace Drupal\nidirect_prisons\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Drupal\Core\Controller\ControllerBase;
+
+class WorldpayNotificationController extends ControllerBase {
+
+  /**
+   * Handle notifications from Worldpay to track if a payment has
+   * succeeded or failed.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The Request.
+   * @return \Symfony\Component\HttpFoundation\Response
+   *   The Response.
+   */
+  public function handleNotification(Request $request) {
```

Comment:
I'm splitting hairs here but there's a few distinct processes happening in the one function that would normally be recommended to be separated, whether just for readability or for improved compatibility for any future unit tests.

There's an authentication process, a transaction process and a final cleanup process in there. Could potentially be reworked so these are encapsulated in their own functions or even classes to give a more robust controller handler here.